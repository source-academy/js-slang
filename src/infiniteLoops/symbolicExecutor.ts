import * as es from 'estree'
import * as stype from './symTypes'

/* Input is asymmetric, i.e. sym + literal
 * or literal + sym, so the idea is to only handle
 * 1 of these cases and reuse it to handle the other.
 * using the 'flipped' flag
 */
function execBinarySymbol(
  node1: stype.SymbolicExecutable,
  node2: stype.SymbolicExecutable,
  op: string,
  flipped: boolean
): stype.SSymbol {
  type opFunction = (value: number, sym: stype.NumberSymbol, flipped: boolean) => stype.SSymbol
  const operators: { [nodeType: string]: opFunction } = {
    '+'(value: number, sym: stype.NumberSymbol, flip: boolean) {
      return { ...sym, constant: sym.constant + value }
    },
    '-'(value: number, sym: stype.NumberSymbol, flip: boolean) {
      if (flip) {
        return { ...sym, constant: value - sym.constant, isPositive: !sym.isPositive }
      } else {
        return { ...sym, constant: sym.constant - value }
      }
    },
    '==='(value: number, sym: stype.NumberSymbol, flip: boolean) {
      return stype.makeInequalitySymbol(sym.name, value - sym.constant, 0)
    },
    '<'(value: number, sym: stype.NumberSymbol, flip: boolean) {
      if (sym.isPositive) {
        if (flip) {
          return stype.makeInequalitySymbol(sym.name, value - sym.constant, 1)
        } else {
          return stype.makeInequalitySymbol(sym.name, value - sym.constant, -1)
        }
      } else {
        const negated = stype.negateNumberSymbol(sym)
        return operators['>'](value, negated, flip)
      }
    },
    '>'(value: number, sym: stype.NumberSymbol, flip: boolean) {
      return operators['<'](value, sym, !flip)
    },
    '<='(value: number, sym: stype.NumberSymbol, flip: boolean) {
      if (flip) {
        return operators['<'](value - 1, sym, flip)
      } else {
        return operators['<'](value + 1, sym, flip)
      }
    },
    '>='(value: number, sym: stype.NumberSymbol, flip: boolean) {
      return operators['<='](value, sym, !flip)
    },
    '!=='(value: number, sym: stype.NumberSymbol, flip: boolean) {
      return stype.negateBooleanSymbol(
        operators['==='](value, sym, false) as stype.InequalitySymbol
      )
    }
  }
  if (node1.type === 'LiteralValueSymbol' && node2.type === 'NumberSymbol') {
    return execBinarySymbol(node2, node1, op, true)
  } else if (node2.type === 'LiteralValueSymbol' && node1.type === 'NumberSymbol') {
    const val = node2.value
    if (typeof val === 'number' && Number.isInteger(val)) {
      const toRun = operators[op]
      if (toRun !== undefined) {
        return toRun(val, node1 as stype.NumberSymbol, flipped)
      }
    }
  } else if (node1.type === 'FunctionSymbol') {
    if (node2.type === 'FunctionSymbol') {
      return stype.makeSequenceSymbol([node1, node2])
    } else {
      return node1
    }
  } else if (node2.type === 'FunctionSymbol') {
    return node2
  }
  return stype.skipSymbol
}

function execLogicalSymbol(
  node1: stype.SymbolicExecutable,
  node2: stype.SymbolicExecutable,
  op: string
): stype.SSymbol {
  if (node1.type === 'LiteralValueSymbol') {
    if (node2.type === 'LiteralValueSymbol') {
      return stype.skipSymbol
    } else {
      return execLogicalSymbol(node2, node1, op)
    }
  } else if (stype.isBooleanSymbol(node1)) {
    if (node2.type === 'LiteralValueSymbol' && typeof node2.value === 'boolean') {
      const val = node2.value
      if ((val && op === '&&') || op === '||') {
        return node1
      }
    } else if (stype.isBooleanSymbol(node2)) {
      return stype.makeLogicalSymbol(node1, node2, op === '&&')
    }
  } else if (node1.type === 'FunctionSymbol') {
    if (node2.type === 'FunctionSymbol') {
      return stype.makeSequenceSymbol([node1, node2])
    } else {
      return node1
    }
  } else if (node2.type === 'FunctionSymbol') {
    return node2
  }
  return stype.skipSymbol
}

function getFromStore(name: string, store: Map<string, stype.SSymbol>[]) {
  for (const st of store) {
    if (st.has(name)) {
      return st.get(name)
    }
  }
  return undefined
}

function ifNullWrapDummyLoc(loc: es.SourceLocation | undefined | null) {
  const pos = { line: 10, column: 1 } as es.Position
  const dummy = { source: null, start: pos, end: pos } as es.SourceLocation
  return loc ? loc : dummy
}

export function getFirstCall(node: es.FunctionDeclaration): stype.FunctionSymbol {
  function doParam(param: es.Node) {
    if (param.type === 'Identifier') {
      return stype.makeNumberSymbol(param.name, 0)
    }
    return stype.unusedSymbol
  }
  const id = node.id as es.Identifier
  const args = node.params.map(doParam)
  return stype.makeFunctionSymbol(id.name, args, ifNullWrapDummyLoc(node.loc))
}

function deleteAllAfterReturn(input: es.Node[]): es.Node[] {
  function helper(nodes: es.Node[]): [boolean, es.Node[]] {
    const output: es.Node[] = []
    for (const nd of nodes) {
      if (nd.type === 'ReturnStatement') {
        output.push(nd)
        return [true, output]
      } else if (nd.type === 'BlockStatement') {
        let newBody
        let changedNodes
        ;[changedNodes, newBody] = helper(nd.body)
        const newBlock = { ...nd, body: newBody } as es.BlockStatement
        output.push(newBlock)
        if (changedNodes) {
          return [true, output]
        }
      } else {
        output.push(nd)
      }
    }
    return [false, output]
  }
  return helper(input)[1]
}

type Executor = (node: es.Node, store: Map<string, stype.SSymbol>[]) => stype.SSymbol

// similar to evaluators in the interpreter
export const nodeToSym: { [nodeType: string]: Executor } = {
  Literal(node: es.Literal, store: Map<string, stype.SSymbol>[]) {
    if (typeof node.value === 'number' || typeof node.value === 'boolean') {
      return stype.makeLiteralValueSymbol(node.value)
    }
    return stype.skipSymbol
  },
  Identifier(node: es.Identifier, store: Map<string, stype.SSymbol>[]) {
    const checkStore = getFromStore(node.name, store)
    if (checkStore) {
      return checkStore
    }
    return stype.skipSymbol
  },
  VariableDeclaration(node: es.VariableDeclaration, store: Map<string, stype.SSymbol>[]) {
    const declaration = node.declarations[0]
    const rhs = declaration.init
    const id = declaration.id as es.Identifier
    if (rhs) {
      const result = symEx(rhs, store)
      if (stype.isTerminal(result)) {
        store[0].set(id.name, result)
        return stype.unusedSymbol
      } else {
        store[0].set(id.name, stype.skipSymbol)
        return result
      }
    }
    return stype.skipSymbol
  },
  FunctionDeclaration(node: es.FunctionDeclaration, store: Map<string, stype.SSymbol>[]) {
    const id = node.id
    if (id?.type === 'Identifier') {
      store[0].set(id.name, stype.skipSymbol)
    }
    return stype.skipSymbol
  },
  ArrowFunctionExpression(node: es.FunctionDeclaration, store: Map<string, stype.SSymbol>[]) {
    return stype.skipSymbol
  },
  ExpressionStatement(node: es.ExpressionStatement, store: Map<string, stype.SSymbol>[]) {
    return symEx(node.expression, store)
  },
  IfStatement(
    node: es.IfStatement | es.ConditionalExpression,
    store: Map<string, stype.SSymbol>[]
  ) {
    const test = symEx(node.test, store)
    const consequent = symEx(node.consequent, store)
    const alternate = node.alternate ? symEx(node.alternate, store) : stype.unusedSymbol
    return stype.makeBranchSymbol(test, consequent, alternate)
  },
  ConditionalExpression(node: es.ConditionalExpression, store: Map<string, stype.SSymbol>[]) {
    return nodeToSym.IfStatement(node, store)
  },
  BlockStatement(node: es.BlockStatement, store: Map<string, stype.SSymbol>[]) {
    const newContext = [new Map()].concat(store)
    const newBody = deleteAllAfterReturn(node.body)
    return stype.makeSequenceSymbol(newBody.map(x => symEx(x, newContext)))
  },
  BinaryExpression(node: es.BinaryExpression, store: Map<string, stype.SSymbol>[]) {
    const lhs = symEx(node.left, store)
    const rhs = symEx(node.right, store)
    return execBinarySymbol(lhs, rhs, node.operator, false)
  },
  UnaryExpression(node: es.UnaryExpression, store: Map<string, stype.SSymbol>[]) {
    const arg = symEx(node.argument, store)
    if (node.operator === '!') {
      if (stype.isBooleanSymbol(arg)) {
        return stype.negateBooleanSymbol(arg)
      } else if (arg.type === 'LiteralValueSymbol') {
        return { ...arg, value: !arg.value }
      }
    } else if (node.operator === '-') {
      if (arg.type === 'NumberSymbol') {
        return stype.negateNumberSymbol(arg)
      } else if (arg.type === 'LiteralValueSymbol') {
        return { ...arg, value: -arg.value }
      }
    }
    return stype.skipSymbol
  },
  LogicalExpression(node: es.LogicalExpression, store: Map<string, stype.SSymbol>[]) {
    const lhs = symEx(node.left, store)
    const rhs = symEx(node.right, store)
    return execLogicalSymbol(lhs, rhs, node.operator)
  },
  CallExpression(node: es.CallExpression, store: Map<string, stype.SSymbol>[]) {
    if (node.callee.type === 'Identifier') {
      const checkShadowed = getFromStore(node.callee.name, store)
      if (checkShadowed?.type === 'SkipSymbol') {
        return stype.skipSymbol
      }
      return stype.makeFunctionSymbol(
        node.callee.name,
        node.arguments.map(x => symEx(x, store)),
        ifNullWrapDummyLoc(node.loc)
      )
    }
    return stype.skipSymbol
  },
  ReturnStatement(node: es.ReturnStatement, store: Map<string, stype.SSymbol>[]) {
    const arg = node.argument
    if (arg === undefined || arg === null || arg.type === 'Identifier' || arg.type === 'Literal') {
      return stype.terminateSymbol
    } else {
      const value = symEx(arg, store)
      if (value.type === 'BranchSymbol') {
        return returnConditional(value)
      }
      return stype.terminalOrSkip(value) ? stype.terminateSymbol : value
    }
  }
}

/* handle the case where we return a conditional,
 * e.g. return true?1:2;
 * we want to turn it into a terminate symbol if
 * we can.
 */
function returnConditional(sym: stype.BranchSymbol): stype.SSymbol {
  let consequent = stype.terminalOrSkip(sym.consequent) ? stype.terminateSymbol : sym.consequent
  let alternate = stype.terminalOrSkip(sym.alternate) ? stype.terminateSymbol : sym.alternate
  if (sym.consequent.type === 'BranchSymbol') {
    consequent = returnConditional(sym.consequent)
  }
  if (sym.alternate.type === 'BranchSymbol') {
    alternate = returnConditional(sym.alternate)
  }

  return stype.makeBranchSymbol(sym.test, consequent, alternate)
}

function symEx(node: stype.SymbolicExecutable, store: Map<string, stype.SSymbol>[]): stype.SSymbol {
  if (stype.isSymbol(node)) {
    return node as stype.SSymbol
  }
  const handler = nodeToSym[node.type]
  if (handler) {
    return handler(node, store)
  }
  return stype.skipSymbol
}

/* initialize the store with function parameters (from firstCall)
 * and globals. e.g. for function fac(x){...}, we create and store an 'x' symbol
 */
function makeStore(
  firstCall: stype.FunctionSymbol,
  constants: [string, number][]
): Map<string, stype.SSymbol>[] {
  const store = [new Map()]

  for (const [name, val] of constants) {
    store[0].set(name, stype.makeLiteralValueSymbol(val))
  }

  for (const v of firstCall.args) {
    if (v.type === 'NumberSymbol') {
      store[0].set(v.name, v)
    }
  }
  return store
}

export function symbolicExecute(node: es.FunctionDeclaration, constants: [string, number][]) {
  const firstCall = getFirstCall(node)
  const store = makeStore(firstCall, constants)
  return symEx(node.body, store)
}
