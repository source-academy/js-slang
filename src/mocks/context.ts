import * as es from 'estree'

import createContext from '../createContext'
import Closure from '../interpreter/closure'
import { Context, Environment, Variant } from '../types'

export function mockContext(chapter = 1, variant: Variant = 'default'): Context {
  return createContext(chapter, variant)
}

export function mockRuntimeContext(): Context {
  const context = createContext()
  context.runtime = {
    break: false,
    debuggerOn: true,
    isRunning: true,
    environments: [],
    nodes: [
      {
        type: 'Literal',
        loc: {
          start: { line: 1, column: 0 },
          end: { line: 1, column: 1 }
        },
        value: 0,
        raw: '0',
        range: [0, 1]
      }
    ]
  }
  return context
}

export function mockClosure(): Closure {
  return new Closure(
    {
      type: 'FunctionExpression',
      loc: null,
      id: null,
      params: [],
      body: {
        type: 'BlockStatement',
        body: []
      }
    } as es.FunctionExpression,
    {} as Environment,
    {} as Context
  )
}
