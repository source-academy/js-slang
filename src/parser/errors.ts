import { Node, SourceLocation } from 'estree'

import { UNKNOWN_LOCATION } from '../constants'
import { ErrorSeverity, ErrorType, SourceError } from '../types'
import { stripIndent } from '../utils/formatters'

export class MissingSemicolonError implements SourceError {
  public type = ErrorType.SYNTAX
  public severity = ErrorSeverity.ERROR
  public constructor(public location: SourceLocation) {}

  public explain() {
    return 'Missing semicolon at the end of statement'
  }

  public elaborate() {
    return 'Every statement must be terminated by a semicolon.'
  }
}

export class TrailingCommaError implements SourceError {
  public type: ErrorType.SYNTAX
  public severity: ErrorSeverity.WARNING
  public constructor(public location: SourceLocation) {}

  public explain() {
    return 'Trailing comma'
  }

  public elaborate() {
    return 'Please remove the trailing comma'
  }
}

export class FatalSyntaxError implements SourceError {
  public type = ErrorType.SYNTAX
  public severity = ErrorSeverity.ERROR
  public constructor(public location: SourceLocation, public message: string) {}

  public explain() {
    return this.message
  }

  public elaborate() {
    return 'There is a syntax error in your program'
  }
}

export class DisallowedConstructError implements SourceError {
  public type = ErrorType.SYNTAX
  public severity = ErrorSeverity.ERROR
  public nodeType: string

  constructor(public node: Node) {
    this.nodeType = this.formatNodeType(this.node.type)
  }

  get location() {
    return this.node.loc ?? UNKNOWN_LOCATION
  }

  public explain() {
    return `${this.nodeType} are not allowed`
  }

  public elaborate() {
    return stripIndent`
        You are trying to use ${this.nodeType}, which is not allowed (yet).
      `
  }

  /**
   * Converts estree node.type into english
   * e.g. ThisExpression -> 'this' expressions
   *      Property -> Properties
   *      EmptyStatement -> Empty Statements
   */
  private formatNodeType(nodeType: string) {
    switch (nodeType) {
      case 'ThisExpression':
        return "'this' expressions"
      case 'Property':
        return 'Properties'
      case 'ImportNamespaceSpecifier':
        return 'Namespace imports'
      default: {
        const words = nodeType.split(/(?=[A-Z])/)
        return words.map((word, i) => (i === 0 ? word : word.toLowerCase())).join(' ') + 's'
      }
    }
  }
}
