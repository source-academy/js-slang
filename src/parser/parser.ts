/* tslint:disable:max-classes-per-file */
import {
  Options as AcornOptions,
  parse as acornParse,
  parseExpressionAt as acornParseAt,
  Position
} from 'acorn'
import { parse as acornLooseParse } from 'acorn-loose'
import { ancestor, AncestorWalkerFn } from 'acorn-walk/dist/walk'
import * as es from 'estree'
import { Context, ErrorSeverity, ErrorType, Rule, SourceError } from '../types'
import { stripIndent } from '../utils/formatters'
import rules from './rules'
import syntaxBlacklist from './syntaxBlacklist'

export class DisallowedConstructError implements SourceError {
  public type = ErrorType.SYNTAX
  public severity = ErrorSeverity.ERROR
  public nodeType: string

  constructor(public node: es.Node) {
    this.nodeType = this.formatNodeType(this.node.type)
  }

  get location() {
    return this.node.loc!
  }

  public explain() {
    return `${this.nodeType} are not allowed`
  }

  public elaborate() {
    return stripIndent`
      You are trying to use ${this.nodeType}, which is not allowed (yet).
    `
  }

  /**
   * Converts estree node.type into english
   * e.g. ThisExpression -> 'this' expressions
   *      Property -> Properties
   *      EmptyStatement -> Empty Statements
   */
  private formatNodeType(nodeType: string) {
    switch (nodeType) {
      case 'ThisExpression':
        return "'this' expressions"
      case 'Property':
        return 'Properties'
      default: {
        const words = nodeType.split(/(?=[A-Z])/)
        return words.map((word, i) => (i === 0 ? word : word.toLowerCase())).join(' ') + 's'
      }
    }
  }
}

export class FatalSyntaxError implements SourceError {
  public type = ErrorType.SYNTAX
  public severity = ErrorSeverity.ERROR
  public constructor(public location: es.SourceLocation, public message: string) {}

  public explain() {
    return this.message
  }

  public elaborate() {
    return 'There is a syntax error in your program'
  }
}

export class MissingSemicolonError implements SourceError {
  public type = ErrorType.SYNTAX
  public severity = ErrorSeverity.ERROR
  public constructor(public location: es.SourceLocation) {}

  public explain() {
    return 'Missing semicolon at the end of statement'
  }

  public elaborate() {
    return 'Every statement must be terminated by a semicolon.'
  }
}

export class TrailingCommaError implements SourceError {
  public type: ErrorType.SYNTAX
  public severity: ErrorSeverity.WARNING
  public constructor(public location: es.SourceLocation) {}

  public explain() {
    return 'Trailing comma'
  }

  public elaborate() {
    return 'Please remove the trailing comma'
  }
}

export function parseAt(source: string, num: number) {
  let theNode: acorn.Node | undefined
  try {
    theNode = acornParseAt(source, num)
  } catch (error) {
    return undefined
  }
  return theNode
}

export function parse(source: string, context: Context, fallbackToLooseParse: boolean = false) {
  let program: es.Program | undefined
  try {
    program = (acornParse(source, createAcornParserOptions(context)) as unknown) as es.Program
    ancestor(program as es.Node, walkers, undefined, context)
  } catch (error) {
    if (error instanceof SyntaxError) {
      // tslint:disable-next-line:no-any
      const loc = (error as any).loc
      const location = {
        start: { line: loc.line, column: loc.column },
        end: { line: loc.line, column: loc.column + 1 }
      }
      context.errors.push(new FatalSyntaxError(location, error.toString()))
    } else {
      throw error
    }
  }
  const hasErrors = context.errors.find(m => m.severity === ErrorSeverity.ERROR)
  if (program && !hasErrors) {
    return program
  } else if (fallbackToLooseParse) {
    return looseParse(source, context)
  } else {
    return undefined
  }
}

const createAcornParserOptions = (context: Context): AcornOptions => ({
  sourceType: 'module',
  ecmaVersion: 6,
  locations: true,
  // tslint:disable-next-line:no-any
  onInsertedSemicolon(end: any, loc: any) {
    context.errors.push(
      new MissingSemicolonError({
        end: { line: loc.line, column: loc.column + 1 },
        start: loc
      })
    )
  },
  // tslint:disable-next-line:no-any
  onTrailingComma(end: any, loc: Position) {
    context.errors.push(
      new TrailingCommaError({
        end: { line: loc.line, column: loc.column + 1 },
        start: loc
      })
    )
  }
})

// Names-extractor needs comments
export function parseForNames(source: string): [es.Program, acorn.Comment[]] {
  let comments: acorn.Comment[] = []
  const options: AcornOptions = {
    sourceType: 'module',
    ecmaVersion: 6,
    locations: true,
    onComment: comments
  }
  let program: es.Program | undefined
  try {
    program = (acornParse(source, options) as unknown) as es.Program
  } catch {
    comments = []
    program = acornLooseParse(source, options)
  }

  return [program, comments]
}

export function looseParse(source: string, context: Context) {
  const program = (acornLooseParse(
    source,
    createAcornParserOptions(context)
  ) as unknown) as es.Program
  return program
}

function createWalkers(
  allowedSyntaxes: { [nodeName: string]: number },
  parserRules: Rule<es.Node>[]
) {
  const newWalkers = new Map<string, AncestorWalkerFn<Context>>()
  const visitedNodes = new Set<es.Node>()

  // Provide callbacks checking for disallowed syntaxes, such as case, switch...
  const syntaxPairs = Object.entries(allowedSyntaxes)
  syntaxPairs.map(pair => {
    const syntax = pair[0]
    newWalkers.set(syntax, (node: es.Node, context: Context, ancestors: [es.Node]) => {
      if (!visitedNodes.has(node)) {
        visitedNodes.add(node)

        if (context.chapter < allowedSyntaxes[node.type]) {
          context.errors.push(new DisallowedConstructError(node))
        }
      }
    })
  })

  // Provide callbacks checking for rule violations, e.g. no block arrow funcs, non-empty lists...
  parserRules.forEach(rule => {
    const checkers = rule.checkers
    const syntaxCheckerPair = Object.entries(checkers)
    syntaxCheckerPair.forEach(pair => {
      const syntax = pair[0]
      const checker = pair[1]
      const oldCheck = newWalkers.get(syntax)!
      const newCheck = (node: es.Node, context: Context, ancestors: es.Node[]) => {
        if (typeof rule.disableOn !== 'undefined' && context.chapter >= rule.disableOn) {
          return
        }
        const errors = checker(node, ancestors)
        errors.forEach(e => context.errors.push(e))
      }
      newWalkers.set(syntax, (node: es.Node, context: Context<any>, ancestors) => {
        oldCheck(node, context, ancestors)
        newCheck(node, context, ancestors)
      })
    })
  })

  return mapToObj(newWalkers)
}

const mapToObj = (map: Map<string, any>) =>
  Array.from(map).reduce((obj, [k, v]) => Object.assign(obj, { [k]: v }), {})

const walkers: { [name: string]: AncestorWalkerFn<Context> } = createWalkers(syntaxBlacklist, rules)
