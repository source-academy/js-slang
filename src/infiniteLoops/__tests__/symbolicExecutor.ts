import { mockContext } from '../../mocks/context'
import { parse } from '../../parser/parser'
import * as es from 'estree'
import { symbolicExecute } from '../symbolicExecutor'

test('sym tree for fib function', () => {
  const code = `
        function fib(x) {
            if (x===0 || x===1) {
                return 1;
            } else {
                return fib(x-1) + fib(x-2);
            }
        }
    `
  const program = parse(code, mockContext())!
  const symTree = symbolicExecute(
    program.body[0] as es.FunctionDeclaration,
    mockContext().runtime.environments[0]
  )
  expect(symTree).toMatchSnapshot()
})

test('sym tree with conditional expr', () => {
  const code = `
        function f(x){return x===1?1:f(x-1);}
    `
  const program = parse(code, mockContext())!
  const symTree = symbolicExecute(
    program.body[0] as es.FunctionDeclaration,
    mockContext().runtime.environments[0]
  )
  expect(symTree).toMatchSnapshot()
})

test('test variable declaration and global consts', () => {
  const code = `
        function f(x) {let j=0;if(x===b){return 1;}else{return x*f(j);}}
    `
  const context = mockContext(4)
  const environment = context.runtime.environments[0]
  const name = 'b'
  const value = 9
  Object.defineProperty(environment.head, name, {
    value,
    writable: true,
    enumerable: true
  })
  const program = parse(code, context)!
  const symTree = symbolicExecute(
    program.body[0] as es.FunctionDeclaration,
    context.runtime.environments[0]
  )
  expect(symTree).toMatchSnapshot()
})

test('test function with no return or other calls', () => {
  const code = `
        function h(x){2;}
    `
  const program = parse(code, mockContext())!
  const symTree = symbolicExecute(
    program.body[0] as es.FunctionDeclaration,
    mockContext().runtime.environments[0]
  )
  expect(symTree).toMatchSnapshot()
})

test('test function with skip symbols', () => {
  const code = `
        function h(x){
            if(g(x)+2) {
                return 2*3;
            } else {
                2+4*5;
            }
        }
    `
  const program = parse(code, mockContext())!
  const symTree = symbolicExecute(
    program.body[0] as es.FunctionDeclaration,
    mockContext().runtime.environments[0]
  )
  expect(symTree).toMatchSnapshot()
})

test('testing inequalities', () => {
  const code = `
        function h(x){
            x<1;1<x;
            x<=1;1<=x;
            x>1;1>x;
            x>=1;1>=x;
            x!==1 && 1!==x;
            x===1 || 1===x;
            !(x<2);
        }
    `
  const program = parse(code, mockContext())!
  const symTree = symbolicExecute(
    program.body[0] as es.FunctionDeclaration,
    mockContext().runtime.environments[0]
  )
  expect(symTree).toMatchSnapshot()
})

test('testing arithmetic', () => {
  const code = `
        function h(x){
            -x; (-1);
            1+x;x+1;
            1-x;x-1;
            const y=-x;
            (-y)+1;
            f(x) + 1;1 + f(x);
        }
    `
  const program = parse(code, mockContext())!
  const symTree = symbolicExecute(
    program.body[0] as es.FunctionDeclaration,
    mockContext().runtime.environments[0]
  )
  expect(symTree).toMatchSnapshot()
})

test('nested conditional return', () => {
  const code = `
          function h(x){
              return x===0 ?
                        x>5 ? 1 : x<2? g(2) : 3 : f(4);
          }
      `
  const program = parse(code, mockContext())!
  const symTree = symbolicExecute(
    program.body[0] as es.FunctionDeclaration,
    mockContext().runtime.environments[0]
  )
  expect(symTree).toMatchSnapshot()
})
