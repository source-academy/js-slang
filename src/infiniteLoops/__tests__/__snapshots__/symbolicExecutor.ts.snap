// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nested conditional return 1`] = `
Object {
  "symbols": Array [
    Object {
      "alternate": Object {
        "args": Array [
          Object {
            "type": "LiteralValueSymbol",
            "value": 4,
          },
        ],
        "name": "f",
        "type": "FunctionSymbol",
      },
      "consequent": Object {
        "alternate": Object {
          "alternate": Object {
            "type": "TerminateSymbol",
          },
          "consequent": Object {
            "args": Array [
              Object {
                "type": "LiteralValueSymbol",
                "value": 2,
              },
            ],
            "name": "g",
            "type": "FunctionSymbol",
          },
          "test": Object {
            "constant": 2,
            "direction": -1,
            "name": "x",
            "type": "InequalitySymbol",
          },
          "type": "BranchSymbol",
        },
        "consequent": Object {
          "type": "TerminateSymbol",
        },
        "test": Object {
          "constant": 5,
          "direction": 1,
          "name": "x",
          "type": "InequalitySymbol",
        },
        "type": "BranchSymbol",
      },
      "test": Object {
        "constant": 0,
        "direction": 0,
        "name": "x",
        "type": "InequalitySymbol",
      },
      "type": "BranchSymbol",
    },
  ],
  "type": "SequenceSymbol",
}
`;

exports[`sym tree for fib function 1`] = `
Object {
  "symbols": Array [
    Object {
      "alternate": Object {
        "symbols": Array [
          Object {
            "symbols": Array [
              Object {
                "args": Array [
                  Object {
                    "constant": -1,
                    "isPositive": true,
                    "name": "x",
                    "type": "NumberSymbol",
                  },
                ],
                "name": "fib",
                "type": "FunctionSymbol",
              },
              Object {
                "args": Array [
                  Object {
                    "constant": -2,
                    "isPositive": true,
                    "name": "x",
                    "type": "NumberSymbol",
                  },
                ],
                "name": "fib",
                "type": "FunctionSymbol",
              },
            ],
            "type": "SequenceSymbol",
          },
        ],
        "type": "SequenceSymbol",
      },
      "consequent": Object {
        "symbols": Array [
          Object {
            "type": "TerminateSymbol",
          },
        ],
        "type": "SequenceSymbol",
      },
      "test": Object {
        "conjunction": false,
        "left": Object {
          "constant": 0,
          "direction": 0,
          "name": "x",
          "type": "InequalitySymbol",
        },
        "right": Object {
          "constant": 1,
          "direction": 0,
          "name": "x",
          "type": "InequalitySymbol",
        },
        "type": "LogicalSymbol",
      },
      "type": "BranchSymbol",
    },
  ],
  "type": "SequenceSymbol",
}
`;

exports[`sym tree with conditional expr 1`] = `
Object {
  "symbols": Array [
    Object {
      "alternate": Object {
        "args": Array [
          Object {
            "constant": -1,
            "isPositive": true,
            "name": "x",
            "type": "NumberSymbol",
          },
        ],
        "name": "f",
        "type": "FunctionSymbol",
      },
      "consequent": Object {
        "type": "TerminateSymbol",
      },
      "test": Object {
        "constant": 1,
        "direction": 0,
        "name": "x",
        "type": "InequalitySymbol",
      },
      "type": "BranchSymbol",
    },
  ],
  "type": "SequenceSymbol",
}
`;

exports[`test function with no return or other calls 1`] = `
Object {
  "symbols": Array [
    Object {
      "type": "LiteralValueSymbol",
      "value": 2,
    },
  ],
  "type": "SequenceSymbol",
}
`;

exports[`test function with skip symbols 1`] = `
Object {
  "symbols": Array [
    Object {
      "alternate": Object {
        "symbols": Array [
          Object {
            "type": "SkipSymbol",
          },
        ],
        "type": "SequenceSymbol",
      },
      "consequent": Object {
        "symbols": Array [
          Object {
            "type": "TerminateSymbol",
          },
        ],
        "type": "SequenceSymbol",
      },
      "test": Object {
        "args": Array [
          Object {
            "constant": 0,
            "isPositive": true,
            "name": "x",
            "type": "NumberSymbol",
          },
        ],
        "name": "g",
        "type": "FunctionSymbol",
      },
      "type": "BranchSymbol",
    },
  ],
  "type": "SequenceSymbol",
}
`;

exports[`test variable declaration and global consts 1`] = `
Object {
  "symbols": Array [
    Object {
      "type": "UnusedSymbol",
    },
    Object {
      "alternate": Object {
        "symbols": Array [
          Object {
            "args": Array [
              Object {
                "type": "SkipSymbol",
              },
            ],
            "name": "f",
            "type": "FunctionSymbol",
          },
        ],
        "type": "SequenceSymbol",
      },
      "consequent": Object {
        "symbols": Array [
          Object {
            "type": "TerminateSymbol",
          },
        ],
        "type": "SequenceSymbol",
      },
      "test": Object {
        "type": "SkipSymbol",
      },
      "type": "BranchSymbol",
    },
  ],
  "type": "SequenceSymbol",
}
`;

exports[`testing arithmetic 1`] = `
Object {
  "symbols": Array [
    Object {
      "constant": -0,
      "isPositive": true,
      "name": "x",
      "type": "NumberSymbol",
    },
    Object {
      "type": "LiteralValueSymbol",
      "value": -1,
    },
    Object {
      "constant": 1,
      "isPositive": true,
      "name": "x",
      "type": "NumberSymbol",
    },
    Object {
      "constant": 1,
      "isPositive": true,
      "name": "x",
      "type": "NumberSymbol",
    },
    Object {
      "constant": 1,
      "isPositive": true,
      "name": "x",
      "type": "NumberSymbol",
    },
    Object {
      "constant": -1,
      "isPositive": true,
      "name": "x",
      "type": "NumberSymbol",
    },
    Object {
      "type": "UnusedSymbol",
    },
    Object {
      "type": "SkipSymbol",
    },
    Object {
      "args": Array [
        Object {
          "constant": 0,
          "isPositive": true,
          "name": "x",
          "type": "NumberSymbol",
        },
      ],
      "name": "f",
      "type": "FunctionSymbol",
    },
    Object {
      "args": Array [
        Object {
          "constant": 0,
          "isPositive": true,
          "name": "x",
          "type": "NumberSymbol",
        },
      ],
      "name": "f",
      "type": "FunctionSymbol",
    },
  ],
  "type": "SequenceSymbol",
}
`;

exports[`testing inequalities 1`] = `
Object {
  "symbols": Array [
    Object {
      "constant": 1,
      "direction": -1,
      "name": "x",
      "type": "InequalitySymbol",
    },
    Object {
      "constant": 1,
      "direction": 1,
      "name": "x",
      "type": "InequalitySymbol",
    },
    Object {
      "constant": 2,
      "direction": -1,
      "name": "x",
      "type": "InequalitySymbol",
    },
    Object {
      "constant": 2,
      "direction": 1,
      "name": "x",
      "type": "InequalitySymbol",
    },
    Object {
      "constant": 1,
      "direction": 1,
      "name": "x",
      "type": "InequalitySymbol",
    },
    Object {
      "constant": 1,
      "direction": -1,
      "name": "x",
      "type": "InequalitySymbol",
    },
    Object {
      "constant": 0,
      "direction": 1,
      "name": "x",
      "type": "InequalitySymbol",
    },
    Object {
      "constant": 0,
      "direction": -1,
      "name": "x",
      "type": "InequalitySymbol",
    },
    Object {
      "conjunction": true,
      "left": Object {
        "conjunction": false,
        "left": Object {
          "constant": 1,
          "direction": 1,
          "name": "x",
          "type": "InequalitySymbol",
        },
        "right": Object {
          "constant": 1,
          "direction": -1,
          "name": "x",
          "type": "InequalitySymbol",
        },
        "type": "LogicalSymbol",
      },
      "right": Object {
        "conjunction": false,
        "left": Object {
          "constant": 1,
          "direction": 1,
          "name": "x",
          "type": "InequalitySymbol",
        },
        "right": Object {
          "constant": 1,
          "direction": -1,
          "name": "x",
          "type": "InequalitySymbol",
        },
        "type": "LogicalSymbol",
      },
      "type": "LogicalSymbol",
    },
    Object {
      "conjunction": false,
      "left": Object {
        "constant": 1,
        "direction": 0,
        "name": "x",
        "type": "InequalitySymbol",
      },
      "right": Object {
        "constant": 1,
        "direction": 0,
        "name": "x",
        "type": "InequalitySymbol",
      },
      "type": "LogicalSymbol",
    },
    Object {
      "constant": 1,
      "direction": 1,
      "name": "x",
      "type": "InequalitySymbol",
    },
  ],
  "type": "SequenceSymbol",
}
`;
